### PR TITLE
Make rebind optional

### PR DESCRIPTION
**Pull request by kanstantsin-chernik**
Proposed changes:
- Optional rebind for shared_factory

Thread/exception safety: safe

Reviewvers:
@krzysztof-jusiak

### CODE REVIEW

File: extension/include/boost/di/extension/injections/shared_factory.hpp [hunk 1]
```diff
@@ -17,7 +17,25 @@
 BOOST_DI_NAMESPACE_BEGIN
 namespace extension {
 
-template <class T, class TFunc>
+struct no_recursion {};
```
Comment (kris-jusiak) [inline extension/include/boost/di/extension/injections/shared_factory.hpp]: Did you consider just use it as a fwd declaration? No need to instansiate that guy, right?

File: extension/include/boost/di/extension/injections/shared_factory.hpp [hunk 2]
```diff
@@ -68,14 +87,14 @@ struct shared_factory_impl {
 #endif
 };
 
-template <class T, class TFunc>
+template <class T, class TPreventRecursion = void, class TFunc>
```
Comment (kris-jusiak) [inline extension/include/boost/di/extension/injections/shared_factory.hpp]: Did you consider reverse the logic here and use true_type, false_type, instead?
```
class TPreventRecursion = std::false_type;
```

File: extension/test/injections/shared_factory.cpp [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #include <cassert>
 
 namespace di = boost::di;
+namespace ext = di::extension;
```
Comment (kris-jusiak) [inline extension/test/injections/shared_factory.cpp:12]: :+1:

File: extension/include/boost/di/extension/injections/shared_factory.hpp (hunk 1 above)
Comment (kanstantsin-chernik) [reply extension/include/boost/di/extension/injections/shared_factory.hpp]: Can you please elaborate a bit?

File: extension/include/boost/di/extension/injections/shared_factory.hpp (hunk 2 above)
Comment (kanstantsin-chernik) [reply extension/include/boost/di/extension/injections/shared_factory.hpp]: Sure, will change

File: extension/include/boost/di/extension/injections/shared_factory.hpp (hunk 1 above)
Comment (kanstantsin-chernik) [reply extension/include/boost/di/extension/injections/shared_factory.hpp]: struct no_recursion;

**Review by kris-jusiak — APPROVED** [review]

**Pull request merged**